libcontainer/cgroups/fscommon/openat2.go

```diff
@@ -0,0 +1,73 @@
+package fscommon
+
+import (
+	"os"
+	"strings"
+	"sync"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	cgroupfsDir    = "/sys/fs/cgroup"
+	cgroupfsPrefix = cgroupfsDir + "/"
+)
+
+var (
+	cgroupFd     int = -1
+	prepOnce     sync.Once
+	prepErr      error
+	resolveFlags uint64
+)
+
+func prepareOpenat2() error {
+	prepOnce.Do(func() {
+		fd, err := unix.Openat2(-1, cgroupfsDir, &unix.OpenHow{Flags: unix.O_DIRECTORY | unix.O_PATH})
+		if err != nil {
+			prepErr = &os.PathError{Op: "openat2", Path: cgroupfsDir, Err: err}
+			return
+		}
+		var st unix.Statfs_t
+		if err = unix.Fstatfs(fd, &st); err != nil {
+			prepErr = &os.PathError{Op: "statfs", Path: cgroupfsDir, Err: err}
+			return
+		}
+		if st.Type == unix.CGROUP2_SUPER_MAGIC {
+			resolveFlags = unix.RESOLVE_NO_XDEV | unix.RESOLVE_NO_SYMLINKS
+		} else {
+			// cgroupfs is a set of separate mounts
+			resolveFlags = unix.RESOLVE_BENEATH
+		}
+	})
+
+	return prepErr
+}
+
+func OpenFile(dir, file string, flags int) (*os.File, error) {
```

Comment:
I do get why you take a `dir` here but the interface just looks a little bit weird, maybe it should be clearer that this is only intended to open cgroup paths...

Reply:
The inital name was `OpenCgroupFile()` but when I saw the other functions in the package are `ReadFile` and `WriteFile` so I renamed accordingly.

Docs is still todo.

Reply:
Fair enough.

Reply:
Oh, and I forgot to mention that existing `ReadFile` and `WriteFile` also have `dir` and `file` arguments (which seems reasonable as all the callers have a cgroup path and want to deal with various files under it) so I'm just modelling after.

Reply:
Yeah since it's common code annoyingly we have to deal with the directory path (including the subsystem name). :/